scipy/sparse/csgraph/_flow.pyx

```diff
@@ -0,0 +1,395 @@
+    if sink < 0 or sink >= csgraph.shape[0]:
+        raise ValueError(f'sink value ({sink}) must be between 0 ' +
+                         f'and {csgraph.shape[0]-1}')
+
+    # Our implementation of Edmonds--Karp assumes that edges always exist
+    # in both directions, so we start by adding the reversed edges whenever
+    # they are missing.
+    m = _add_reverse_edges(csgraph)
+    rev_edge_ptr, tails = _make_edge_pointers(m)
+    residual = _edmonds_karp(m.indptr, tails, m.indices,
+                             m.data, rev_edge_ptr, source, sink)
+    residual_array = np.asarray(residual)
+    residual_matrix = csr_matrix((residual_array, m.indices, m.indptr),
+                                 shape=m.shape)
+    source_flow = residual_array[m.indptr[source]:m.indptr[source + 1]]
+    return MaximumFlowResult(source_flow.sum(), residual_matrix)
+
+
+def _add_reverse_edges(a):
+    """Add reversed edges to all edges in a graph.
+
+    This adds to a given directed weighted graph all edges in the reverse
+    direction and give them weight 0, unless they already exist.
+
+    Parameters
+    ----------
+    a : csr_matrix
+        The square matrix in CSR format representing a directed graph
+
+    Returns
+    -------
+    res : csr_matrix
+        A new matrix in CSR format in which the missing edges are represented
+        by explicit zeros.
+
+    """
+    # This is equivalent to just looping over all (i, j) and letting
+    # the weight of (j, i) be zero if it is not already defined.
+    # Using NumPy to do so is several orders of magnitude faster
+    # than the naive solution. We first create create the sorted
+    # list {i*n + j} for nonzero (i, j), then use binary search
+    # to determine if j*n + i belongs to the list. To ensure that
+    # the haystack below is sorted, we presort the column indices if
+    # necessary.
+    if not a.has_sorted_indices:
+        a = a.sorted_indices()
+    acoo = a.tocoo()
+    rows = acoo.row
+    cols = acoo.col
+    n = a.shape[0]
+    haystack = rows*n + cols
+    needles = cols*n + rows
+    indices = np.searchsorted(haystack, needles)
+    # As searchsorted returns a.nnz when a given needle is
+    # larger than the largest element of the haystack, we
+    # explicitly add an element not in needles at the end.
+    haystack = np.concatenate([haystack, [-1]])
+    indexed = haystack[indices]
+    diff = indexed != needles
+    # At this point, diff contains the indices of the (j, i)
+    # to be added. We make use of the COO initializer for CSR
+    # matrices which ensures that 0s are added explicitly.
+    indices_to_add = np.where(diff)[0]
+    data_to_add = np.zeros((indices_to_add.shape[0],), dtype=np.int32)
+    acoo.row = np.concatenate([acoo.row, cols[indices_to_add]])
+    acoo.col = np.concatenate([acoo.col, rows[indices_to_add]])
+    acoo.data = np.concatenate([acoo.data, data_to_add])
+    # Repeat process to get reverse pointers
```

Comment:
This comment is a little out of place, not immediately clear to me what is meant

Reply:
Good catch. In an earlier version, the functionality of `_add_reverse_edges` was part of `_make_edge_pointers`; at some point, I extracted the method, making this comment outdated. Will remove.